File: python/most_frequently_used_words_in_a_text.py
# https://www.codewars.com/kata/51e056fe544cf36c410000fb

import re
from collections import Counter

def top_3_words(text):
    words = re.findall(r"[a-z']*[a-z][a-z']*", text.lower())
    word_counts = Counter(words)
    return [word for word, count in word_counts.most_common(3)]

print(top_3_words("e e e e DDD ddd DdD: ddd ddd aa aA Aa, bb cc cC e e e"))

File: js/most_frequently_used_words_in_a_text.js
// https://www.codewars.com/kata/51e056fe544cf36c410000fb

function topThreeWords(text) {
    if(!text) {
        return []
    }
    
    const words = text.toLowerCase().match(/[a-z']*[a-z][a-z']*/g);
    const wordCounts = {};
    
    if (!words) {
        return []
    }

    words.forEach(word => {
        wordCounts[word] = (wordCounts[word] || 0) + 1;
    });

    const sortedWords = Object.keys(wordCounts).sort((a, b) => wordCounts[b] - wordCounts[a]);
    return sortedWords.slice(0, 3);
}

console.log(topThreeWords("e e e e DDD ddd DdD: ddd ddd aa aA Aa, bb cc cC e e e"));
